core/src/main/java/com/linecorp/armeria/common/stream/AbstractMulticastStreamFactory.java

```
@@ -0,0 +1,596 @@
+                }
+                notifyDownstreams();
+            });
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            upstreamSubscription = s;
+            downstreamSubscriptions.forEach(downstream -> {
+                if (downstream.updateOnSubscribe()) {
+                    if (downstream.executor() != null) {
+                        downstream.executor().execute(() -> downstream.subscriber().onSubscribe(downstream));
+                    } else {
+                        downstream.subscriber().onSubscribe(downstream);
+                    }
+                }
+            });
+            if (state == State.CLOSED) {
+                s.cancel();
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (upstreamOffset >= Integer.MAX_VALUE) {
+                //TODO(minwoox) limit by the size of data not by the size of contentList.
+                upstreamSubscription.cancel();
+                throw new IllegalStateException("Published numbers of data is greater than Integer.MAX_VALUE");
+            }
+            contentList.add(t);
+            upstreamOffset++;
+            notifyDownstreams();
+        }
+
+        /**
+         * Handled by {@link #notifyDownstreamWhenUpstreamCompleted(StreamMessage)} instead.
+         */
+        @Override
+        public void onError(Throwable t) {}
+
+        /**
+         * Handled by {@link #notifyDownstreamWhenUpstreamCompleted(StreamMessage)} instead.
+         */
+        @Override
+        public void onComplete() {}
+
+        void subscribe(DownstreamSubscription subscription) {
+            if (state == State.OPEN) {
+                downstreamSubscriptions.add(subscription);
+                if (state == State.CLOSED) {
+                    downstreamSubscriptions.remove(subscription);
+                    throw new IllegalStateException("This factory has been closed.");
+                }
+            }
```

- Should throw an exception in the `else` block as well.
- `"factory closed already"`